rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -364,54 +400,144 @@ __declare_parameter_common(
   return result;
 }
 
+static
 const rclcpp::ParameterValue &
-NodeParameters::declare_parameter(
+declare_parameter_helper(
   const std::string & name,
+  rclcpp::ParameterType type,
   const rclcpp::ParameterValue & default_value,
-  const rcl_interfaces::msg::ParameterDescriptor & parameter_descriptor,
-  bool ignore_override)
+  rcl_interfaces::msg::ParameterDescriptor parameter_descriptor,
+  bool ignore_override,
+  std::map<std::string, rclcpp::node_interfaces::ParameterInfo> & parameters,
+  const std::map<std::string, rclcpp::ParameterValue> & overrides,
+  CallbacksContainerType & callback_container,
+  const OnParametersSetCallbackType & callback,
+  rclcpp::Publisher<rcl_interfaces::msg::ParameterEvent> * events_publisher,
+  const std::string & combined_name,
+  rclcpp::node_interfaces::NodeClockInterface & node_clock)
 {
-  std::lock_guard<std::recursive_mutex> lock(mutex_);
-
-  ParameterMutationRecursionGuard guard(parameter_modification_enabled_);
-
   // TODO(sloretz) parameter name validation
   if (name.empty()) {
     throw rclcpp::exceptions::InvalidParametersException("parameter name must not be empty");
   }
 
   // Error if this parameter has already been declared and is different
-  if (__lockless_has_parameter(parameters_, name)) {
+  if (__lockless_has_parameter(parameters, name)) {
     throw rclcpp::exceptions::ParameterAlreadyDeclaredException(
             "parameter '" + name + "' has already been declared");
   }
 
+  if (!parameter_descriptor.dynamic_typing) {
+    if (rclcpp::PARAMETER_NOT_SET == type) {
+      type = default_value.get_type();
+    }
+    if (rclcpp::PARAMETER_NOT_SET == type) {
+      throw rclcpp::exceptions::InvalidParameterTypeException{
+              name,
+              "cannot declare a statically typed parameter with an uninitialized value"
+      };
+    }
+    parameter_descriptor.type = static_cast<uint8_t>(type);
+  }
+
   rcl_interfaces::msg::ParameterEvent parameter_event;
   auto result = __declare_parameter_common(
     name,
     default_value,
     parameter_descriptor,
-    parameters_,
-    parameter_overrides_,
-    on_parameters_set_callback_container_,
-    on_parameters_set_callback_,
+    parameters,
+    overrides,
+    callback_container,
+    callback,
     &parameter_event,
     ignore_override);
 
   // If it failed to be set, then throw an exception.
   if (!result.successful) {
+    constexpr const char type_error_msg_start[] = "Wrong parameter type";
+    if (
+      0u == std::strncmp(
+        result.reason.c_str(), type_error_msg_start, sizeof(type_error_msg_start) - 1))
+    {
+      // TODO(ivanpauno): Refactor the logic so we don't need the above `strncmp` and we can
```

Comment:
I don't have any immediate thoughts about how to do this, so I guess it's okay to do in a follow-up PR.

If you don't plan to tackle it in the near-term, can you create an issue for visibility after this PR is merged?

Reply:
> If you don't plan to tackle it in the near-term, can you create an issue for visibility after this PR is merged?

Sure, I think the current code is okayish though.